lib/vector-core/src/enrichment_table/enrichment_tables.rs

```diff
@@ -0,0 +1,286 @@
+/// the `EnrichmentTable` trait. Next, the tables are passed through Vectors `Transform` components,
+/// particularly the `Remap` transform. These Transforms are able to determine which fields we will
+/// want to lookup whilst Vector is running. They can notify the tables of these fields no that the
+/// data can be indexed.
+///
+/// During this phase, the data needs to be mutated and shared around a number of potential
+/// different components. Also, performance is not absolutely critical at this phase. Thus, the
+/// data is stored within a `Mutex` - stored in the `loading` field of this struct.
+///
+/// ## 2. Reading
+///
+/// Once all the data has been loaded we can move to the next stage. This is signified by calling
+/// the `finish_load` method. At this point all the data is swapped out of the `Mutex` in the loading
+/// field into the `ArcSwap` of the `tables` field. `ArcSwap` provides lock-free read-only access
+/// to the data. From this point on we have fast, efficient read-only access and can no longer add
+/// indexes or otherwise mutate the data .
+///
+/// Cloning this object is designed to be cheap. The underlying data will be shared by all clones.
+///
+#[derive(Clone, Default)]
+pub struct EnrichmentTables {
+    loading: Arc<Mutex<Option<HashMap<String, Box<dyn EnrichmentTable + Send + Sync>>>>>,
+    tables: Arc<ArcSwap<Option<HashMap<String, Box<dyn EnrichmentTable + Send + Sync>>>>>,
+}
+
+impl EnrichmentTables {
+    pub fn new(tables: HashMap<String, Box<dyn EnrichmentTable + Send + Sync>>) -> Self {
+        Self {
+            loading: Arc::new(Mutex::new(Some(tables))),
+            tables: Arc::new(ArcSwap::default()),
+        }
+    }
+
+    /// Swap the data out of the `Mutex` into the `ArcSwap`.
+    /// From this point we can no longer add indexes to the tables, but are now allowed to read the
+    /// data.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the Mutex is poisoned.
+    pub fn finish_load(&self) {
+        let mut tables_lock = self.loading.lock().unwrap();
+        let tables = tables_lock.take();
+        self.tables.swap(Arc::new(tables));
+    }
+}
+
+impl std::fmt::Debug for EnrichmentTables {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let tables = self.tables.load();
+        match **tables {
+            Some(ref tables) => {
+                let mut tables = tables.iter().fold(String::from("("), |mut s, (key, _)| {
+                    s.push_str(key);
+                    s.push_str(", ");
+                    s
+                });
+
+                tables.truncate(std::cmp::max(tables.len(), 0));
+                tables.push(')');
+
+                write!(f, "EnrichmentTables {}", tables)
+            }
+            None => write!(f, "EnrichmentTables loading"),
+        }
+    }
+}
+
+#[cfg(feature = "vrl")]
+impl vrl_core::EnrichmentTables for EnrichmentTables {
```

Comment:
Full disclosure, with each of the functions in the impl (`get_tables`, `find_table_row` and `add_index`) I had them all consistently doing:

```
match ... { 
  None => ...
  Some(..) => ...
}
```

But clippy has made me do a different thing for each function, which honestly doesn't make much sense to me.